cpg-console Binary cannot be used

Hello, I used jdk17 to compile the source code of cpg8.0.0. The compilation was successful, but the final cpg-console binary cannot be used. How should I solve it?


#jdk environment：
  [root@LIN-3580BCC1C73 cpg-8.0.0]# java -version
   openjdk version "17.0.4.1" 2022-08-12
   OpenJDK Runtime Environment Temurin-17.0.4.1+1 (build 17.0.4.1+1)
  OpenJDK 64-Bit Server VM Temurin-17.0.4.1+1 (build 17.0.4.1+1, mixed mode, sharing)

#log
  [root@LIN-3580BCC1C73 cpg-8.0.0]# ./gradlew installDist
Type-safe project accessors is an incubating feature.

> Configure project :
Java frontend is enabled
C/C++ frontend is enabled
Go frontend is enabled
Python frontend is enabled
LLVM frontend is disabled
TypeScript frontend is enabled
Ruby frontend is enabled

> Task :cpg-language-go:compileGolang
go: downloading github.com/mattn/go-pointer v0.0.1
lib/cpg/main.go:44:2: github.com/mattn/go-pointer@v0.0.1: go-pointer/@v/v0.0.1.zip: 401 Unauthorized

> Task :cpg-language-typescript:npmBuild

> bundle
> rollup -c


src/parser.ts → ../../../build/resources/main/nodejs...

> Task :cpg-core:compileKotlin
w: file:///media/vdc/software/cpg-8.0.0/cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/TranslationManager.kt:328:15 Unchecked cast: Any to MutableMap<String, String>
w: file:///media/vdc/software/cpg-8.0.0/cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/CompilationDatabase.kt:303:17 Name shadowed: isLibrary
w: file:///media/vdc/software/cpg-8.0.0/cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/graph/DeclarationHolder.kt:70:49 Unchecked cast: DeclarationHolder to T
w: file:///media/vdc/software/cpg-8.0.0/cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/graph/Extensions.kt:606:17 Unchecked cast: Expression to N
w: file:///media/vdc/software/cpg-8.0.0/cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/graph/Extensions.kt:613:7 Unchecked cast: List<Node> to List<CallExpression>
w: file:///media/vdc/software/cpg-8.0.0/cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/graph/NodeBuilder.kt:107:19 Unchecked cast: MetadataProvider? to CodeAndLocationProvider<Any>
w: file:///media/vdc/software/cpg-8.0.0/cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/graph/StatementHolder.kt:79:17 Name shadowed: before
w: file:///media/vdc/software/cpg-8.0.0/cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/graph/edge/PropertyEdge.kt:180:48 Unchecked cast: PropertyEdge<out Node> to PropertyEdge<T>
w: file:///media/vdc/software/cpg-8.0.0/cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/graph/edge/PropertyEdge.kt:199:72 Unchecked cast: Node to T
w: file:///media/vdc/software/cpg-8.0.0/cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/graph/edge/PropertyEdge.kt:220:29 Unchecked cast: Collection<*>! to MutableCollection<Node>
w: file:///media/vdc/software/cpg-8.0.0/cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/graph/statements/ForEachStatement.kt:65:13 Parameter 'value' is never used
w: file:///media/vdc/software/cpg-8.0.0/cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/helpers/CommentMatcher.kt:151:17 Name shadowed: closestLine
w: file:///media/vdc/software/cpg-8.0.0/cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/helpers/SubgraphWalker.kt:167:42 Unchecked cast: Any to List<PropertyEdge<Node>>
w: file:///media/vdc/software/cpg-8.0.0/cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/helpers/SubgraphWalker.kt:174:49 Unchecked cast: Any to Collection<Node>
w: file:///media/vdc/software/cpg-8.0.0/cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/helpers/Util.kt:145:5 Expected performance impact from inlining is insignificant. Inlining works best for functions with parameters of functional types
w: file:///media/vdc/software/cpg-8.0.0/cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/helpers/Util.kt:162:5 Expected performance impact from inlining is insignificant. Inlining works best for functions with parameters of functional types
w: file:///media/vdc/software/cpg-8.0.0/cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/helpers/Util.kt:179:5 Expected performance impact from inlining is insignificant. Inlining works best for functions with parameters of functional types
w: file:///media/vdc/software/cpg-8.0.0/cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/helpers/Util.kt:191:5 Expected performance impact from inlining is insignificant. Inlining works best for functions with parameters of functional types
w: file:///media/vdc/software/cpg-8.0.0/cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/helpers/Util.kt:203:5 Expected performance impact from inlining is insignificant. Inlining works best for functions with parameters of functional types
w: file:///media/vdc/software/cpg-8.0.0/cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/CXXCallResolverHelper.kt:268:5 Parameter 'ctx' is never used
w: file:///media/vdc/software/cpg-8.0.0/cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/ControlDependenceGraphPass.kt:180:5 Parameter 'currentWorklist' is never used
w: file:///media/vdc/software/cpg-8.0.0/cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/EvaluationOrderGraphPass.kt:794:28 Parameter 'loopStatement' is never used
w: file:///media/vdc/software/cpg-8.0.0/cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/Pass.kt:129:60 Unchecked cast: PassConfiguration? to T
w: file:///media/vdc/software/cpg-8.0.0/cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/StatisticsCollectionPass.kt:54:27 'translationUnits: List<TranslationUnitDeclaration>' is deprecated. translation units of individual components should be accessed instead
w: file:///media/vdc/software/cpg-8.0.0/cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/processing/strategy/Strategy.kt:40:21 Parameter 'x' is never used

> Task :cpg-language-typescript:npmBuild
created ../../../build/resources/main/nodejs in 23.7s

> Task :cpg-analysis:compileKotlin
w: file:///media/vdc/software/cpg-8.0.0/cpg-analysis/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/UnreachableEOGPass.kt:95:5 Parameter 'currentWorklist' is never used
w: file:///media/vdc/software/cpg-8.0.0/cpg-analysis/src/main/kotlin/de/fraunhofer/aisec/cpg/query/QueryTree.kt:66:69 Conditional branch result of type String is implicitly cast to Any?
w: file:///media/vdc/software/cpg-8.0.0/cpg-analysis/src/main/kotlin/de/fraunhofer/aisec/cpg/query/QueryTree.kt:66:121 Conditional branch result of type T is implicitly cast to Any?
w: file:///media/vdc/software/cpg-8.0.0/cpg-analysis/src/main/kotlin/de/fraunhofer/aisec/cpg/query/QueryTree.kt:148:32 Unchecked cast: T to Comparable<Any>

> Task :cpg-language-java:compileKotlin
w: file:///media/vdc/software/cpg-8.0.0/cpg-language-java/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/java/DeclarationHandler.kt:381:9 Parameter 'annotationConstDecl' is never used
w: file:///media/vdc/software/cpg-8.0.0/cpg-language-java/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/java/DeclarationHandler.kt:390:9 Parameter 'annotationMemberDecl' is never used
w: file:///media/vdc/software/cpg-8.0.0/cpg-language-java/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/java/ExpressionHandler.kt:432:13 Variable 'value' is never used
w: file:///media/vdc/software/cpg-8.0.0/cpg-language-java/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/java/StatementHandler.kt:296:38 Parameter 'stmt' is never used
w: file:///media/vdc/software/cpg-8.0.0/cpg-language-java/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/java/StatementHandler.kt:399:13 Name shadowed: token
w: file:///media/vdc/software/cpg-8.0.0/cpg-language-java/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/java/StatementHandler.kt:409:13 Name shadowed: token
w: file:///media/vdc/software/cpg-8.0.0/cpg-language-java/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/java/StatementHandler.kt:555:40 Name shadowed: stmt

> Task :cpg-language-go:compileKotlin
w: file:///media/vdc/software/cpg-8.0.0/cpg-language-go/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/golang/GoLanguageFrontend.kt:220:13 Name shadowed: type

> Task :cpg-language-cxx:compileKotlin
w: file:///media/vdc/software/cpg-8.0.0/cpg-language-cxx/src/main/kotlin/de/fraunhofer/aisec/cpg/frontends/cxx/DeclarationHandler.kt:623:9 Parameter 'ctx' is never used
w: file:///media/vdc/software/cpg-8.0.0/cpg-language-cxx/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/CXXExtraPass.kt:60:48 Parameter 'currClass' is never used

> Task :cpg-neo4j:compileKotlin
w: file:///media/vdc/software/cpg-8.0.0/cpg-neo4j/src/main/kotlin/de/fraunhofer/aisec/cpg_vis_neo4j/Application.kt:409:26 'autoIndex(String!): Configuration.Builder!' is deprecated. Deprecated in Java
w: file:///media/vdc/software/cpg-8.0.0/cpg-neo4j/src/main/kotlin/de/fraunhofer/aisec/cpg_vis_neo4j/Application.kt:495:52 Unchecked cast: KClass<out Any!> to KClass<out Pass<*>>

> Task :cpg-console:compileKotlin
w: file:///media/vdc/software/cpg-8.0.0/cpg-console/src/main/kotlin/de/fraunhofer/aisec/cpg/analysis/NullPointerCheck.kt:52:27 'translationUnits: List<TranslationUnitDeclaration>' is deprecated. translation units of individual components should be accessed instead
w: file:///media/vdc/software/cpg-8.0.0/cpg-console/src/main/kotlin/de/fraunhofer/aisec/cpg/analysis/OutOfBoundsCheck.kt:52:27 'translationUnits: List<TranslationUnitDeclaration>' is deprecated. translation units of individual components should be accessed instead

Deprecated Gradle features were used in this build, making it incompatible with Gradle 9.0.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/8.5/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

BUILD SUCCESSFUL in 56s
82 actionable tasks: 71 executed, 11 up-to-date


#cpg-console error

 [root@LIN-3580BCC1C73 cpg-8.0.0]# cpg-console/build/install/cpg-console/bin/cpg-console
ki-shell 0.5.2/1.9.0
type :h for help
[0] :tr cpg-language-java/src/test/resources/Issue285.java
ERROR Class 'de.fraunhofer.aisec.cpg.TranslationConfiguration' was compiled with an incompatible version of Kotlin. The binary version of its metadata is 1.9.0, expected version is 1.7.1.
The class is loaded from /media/vdc/software/cpg-8.0.0/cpg-console/build/install/cpg-console/lib/cpg-core.jar!/de/fraunhofer/aisec/cpg/TranslationConfiguration.class (Line_1.kts:1:32)
ERROR Class 'de.fraunhofer.aisec.cpg.TranslationManager' was compiled with an incompatible version of Kotlin. The binary version of its metadata is 1.9.0, expected version is 1.7.1.
The class is loaded from /media/vdc/software/cpg-8.0.0/cpg-console/build/install/cpg-console/lib/cpg-core.jar!/de/fraunhofer/aisec/cpg/TranslationManager.class (Line_2.kts:1:32)
ERROR Unresolved reference: evaluate (Line_7.kts:1:38)
ERROR Unresolved reference: allChildren (Line_8.kts:1:38)
ERROR Unresolved reference: ast (Line_9.kts:1:38)
ERROR Unresolved reference: dfgFrom (Line_10.kts:1:38)
ERROR Unresolved reference: byName (Line_11.kts:1:38)
ERROR Unresolved reference: body (Line_12.kts:1:38)
ERROR Unresolved reference: capacity (Line_13.kts:1:38)
ERROR Unresolved reference: printCode (Line_14.kts:1:40)
ERROR Unresolved reference: TranslationConfiguration (Line_16.kts:2:17)
ERROR Unresolved reference: TranslationManager (Line_17.kts:1:16)
ERROR Unresolved reference: config (Line_17.kts:1:52)
ERROR Unresolved reference: analyzer (Line_18.kts:1:14)
ERROR Unresolved reference: result (Line_19.kts:1:10)

--- a/buildSrc/src/main/kotlin/cpg.common-conventions.gradle.kts
+++ b/buildSrc/src/main/kotlin/cpg.common-conventions.gradle.kts
@@ -97,9 +97,7 @@ signing {
 //
 // specify Java & Kotlin JVM version
 kotlin {
-    jvmToolchain {
-        languageVersion.set(JavaLanguageVersion.of(17))
-    }
+    jvmToolchain(17)
 }
 
 tasks.withType<KotlinCompile> {

--- a/cpg-console/build.gradle.kts
+++ b/cpg-console/build.gradle.kts
@@ -52,12 +52,25 @@ tasks.withType<Test> {
     }
 }
 
+
+repositories {
+    maven {
+        setUrl("https://jitpack.io")
+    }
+}
+
 dependencies {
     // CPG
     api(projects.cpgAnalysis)
     api(projects.cpgNeo4j)
 
-    implementation(libs.kotlin.script.runtime)
-    implementation(libs.kotlin.coroutines.core)
     implementation(libs.kotlin.ki.shell)
+
+    // ki-shell dependencies, that we need to specify manually because we are using the JitPack "release"
+    implementation(libs.kotlin.script.runtime)
+    implementation(libs.bundles.kotlin.scripting)
+    implementation(libs.antlr.runtime)
+    implementation(libs.kotlin.coroutines.core)
+    implementation(libs.jline)
+
 }

--- a/cpg-console/src/main/kotlin/de/fraunhofer/aisec/cpg/console/CpgConsole.kt
+++ b/cpg-console/src/main/kotlin/de/fraunhofer/aisec/cpg/console/CpgConsole.kt
@@ -29,11 +29,11 @@ import de.fraunhofer.aisec.cpg.analysis.MultiLineToStringStyle
 import de.fraunhofer.aisec.cpg.graph.Node
 import kotlin.script.experimental.api.ScriptCompilationConfiguration
 import kotlin.script.experimental.api.ScriptEvaluationConfiguration
+import kotlin.script.experimental.api.compilerOptions
 import kotlin.script.experimental.jvm.baseClassLoader
 import kotlin.script.experimental.jvm.defaultJvmScriptingHostConfiguration
 import kotlin.script.experimental.jvm.dependenciesFromClassloader
 import kotlin.script.experimental.jvm.jvm
-import org.jetbrains.kotlinx.ki.shell.KotlinShell
 import org.jetbrains.kotlinx.ki.shell.Plugin
 import org.jetbrains.kotlinx.ki.shell.Shell
 import org.jetbrains.kotlinx.ki.shell.configuration.CachedInstance
@@ -52,13 +52,14 @@ object CpgConsole {
                 ScriptCompilationConfiguration {
                     jvm {
                         dependenciesFromClassloader(
-                            classLoader = KotlinShell::class.java.classLoader,
+                            classLoader = CpgConsole::class.java.classLoader,
                             wholeClasspath = true
                         )
                     }
+                    compilerOptions("-jvm-target=17") // this needs to match the JVM toolchain target in the CPG
                 },
                 ScriptEvaluationConfiguration {
-                    jvm { baseClassLoader(Shell::class.java.classLoader) }
+                    jvm { baseClassLoader(CpgConsole::class.java.classLoader) }
                 }
             )
 

--- a/cpg-console/src/main/kotlin/de/fraunhofer/aisec/cpg/console/CpgConsole.kt
+++ b/cpg-console/src/main/kotlin/de/fraunhofer/aisec/cpg/console/CpgConsole.kt
@@ -56,7 +56,9 @@ object CpgConsole {
                             wholeClasspath = true
                         )
                     }
-                    compilerOptions("-jvm-target=17") // this needs to match the JVM toolchain target in the CPG
+                    compilerOptions(
+                        "-jvm-target=17"
+                    ) // this needs to match the JVM toolchain target in the CPG
                 },
                 ScriptEvaluationConfiguration {
                     jvm { baseClassLoader(CpgConsole::class.java.classLoader) }

--- a/cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/graph/Extensions.kt
+++ b/cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/graph/Extensions.kt
@@ -527,6 +527,10 @@ fun Node.followPrevDFG(predicate: (Node) -> Boolean): MutableList<Node>? {
     return null
 }
 
+/** Returns all [Node] children in this graph, starting with this [Node]. */
+val Node?.nodes: List<Node>
+    get() = this.allChildren()
+
 /** Returns all [CallExpression] children in this graph, starting with this [Node]. */
 val Node?.calls: List<CallExpression>
     get() = this.allChildren()

--- a/gradle/libs.versions.toml
+++ b/gradle/libs.versions.toml
@@ -9,12 +9,17 @@ nexus-publish = "1.3.0"
 [libraries]
 kotlin-reflect = { module = "org.jetbrains.kotlin:kotlin-reflect", version.ref = "kotlin"}
 kotlin-script-runtime = { module = "org.jetbrains.kotlin:kotlin-script-runtime", version.ref = "kotlin"}
+kotlin-scripting-common = { module = "org.jetbrains.kotlin:kotlin-scripting-common", version.ref = "kotlin"}
+kotlin-scripting-jvm = { module = "org.jetbrains.kotlin:kotlin-scripting-jvm", version.ref = "kotlin"}
+kotlin-scripting-dependencies = { module = "org.jetbrains.kotlin:kotlin-scripting-dependencies", version.ref = "kotlin"}
+kotlin-scripting-dependencies-maven-all = { module = "org.jetbrains.kotlin:kotlin-scripting-dependencies-maven-all", version.ref = "kotlin"}
+kotlin-scripting-ide-services = { module = "org.jetbrains.kotlin:kotlin-scripting-ide-services", version.ref = "kotlin"}
 kotlin-coroutines-core = { module = "org.jetbrains.kotlinx:kotlinx-coroutines-core", version = "1.8.0"}
-kotlin-ki-shell = { module = "org.jetbrains.kotlinx:ki-shell", version = "0.5.2"}
+kotlin-ki-shell = { module = "com.github.Kotlin:kotlin-interactive-shell", version = "5b1ff4d821"}
 kotlin-test-junit5 = { module = "org.jetbrains.kotlin:kotlin-test-junit5", version.ref = "kotlin"}  # this is only needed for the testFixtures in cpg-core, everywhere else kotlin("test") is used
 
 log4j-impl = { module = "org.apache.logging.log4j:log4j-slf4j2-impl", version.ref = "log4j" }
-log4j-core = { module= "org.apache.logging.log4j:log4j-core", version.ref = "log4j" }
+log4j-core = { module = "org.apache.logging.log4j:log4j-core", version.ref = "log4j" }
 
 apache-commons-lang3 = { module = "org.apache.commons:commons-lang3", version = "3.14.0"}
 neo4j-ogm-core = { module = "org.neo4j:neo4j-ogm-core", version.ref = "neo4j"}
@@ -30,7 +35,9 @@ picocli = { module = "info.picocli:picocli", version = "4.7.0"}
 picocli-codegen = { module = "info.picocli:picocli-codegen", version = "4.7.0"}
 jep = { module = "black.ninia:jep", version = "4.2.0" }  # build.yml uses grep to extract the jep verison number for CI/CD purposes
 llvm = { module = "org.bytedeco:llvm-platform", version = "16.0.4-1.5.9"}
-jruby = {module = "org.jruby:jruby-core", version = "9.4.3.0" }
+jruby = { module = "org.jruby:jruby-core", version = "9.4.3.0" }
+jline = { module = "org.jline:jline", version = "3.25.1" }
+antlr-runtime = { module = "org.antlr:antlr4-runtime", version = "4.8-1" } # we cannot upgrade until ki-shell upgrades this!
 
 # test
 junit-params = { module = "org.junit.jupiter:junit-jupiter-params", version = "5.10.0"}
@@ -47,6 +54,13 @@ nexus-publish-gradle = { module = "io.github.gradle-nexus:publish-plugin", versi
 [bundles]
 log4j = ["log4j-impl", "log4j-core"]
 neo4j = ["neo4j-ogm-core", "neo4j-ogm-bolt-driver"]
+kotlin-scripting = [
+    "kotlin-scripting-common",
+    "kotlin-scripting-jvm",
+    "kotlin-scripting-ide-services",
+    "kotlin-scripting-dependencies",
+    "kotlin-scripting-dependencies-maven-all",
+]
 
 [plugins]
 kotlin = { id = "org.jetbrains.kotlin.jvm", version.ref = "kotlin"}

